brian2cuda/cuda_generator.py

```diff
@@ -566,9 +586,9 @@ def _add_user_function(self, varname, variable):
                     type_str = self.c_data_type(ns_value.dtype) + '*'
                 else:  # e.g. a function
                     type_str = 'py::object'
-                support_code.append('static {0} _namespace{1};'.format(type_str,
-                                                                       ns_key))
-                pointers.append('_namespace{0} = {1};'.format(ns_key, ns_key))
+                #support_code.append('__device__ {0} _namespace{1};'.format(type_str,
+                #                                                           ns_key))
+                #pointers.append('_namespace{0} = {1};'.format(ns_key, ns_key))
```

Comment:
While implementing TimedArray, I came across some code that has been updated in Brian like 5 years ago but never made it into Brian2CUDA 😅. So I've been updating some things and came across this section here on function namespaces (guess this changed in brian-team/brian2#380). But I don't think I actually understand what the use case is where I need these namespaces. @mstimberg Could you shed some light here? :) 

This function is copied from `CPPCodeGenerator` to `CUDACodeGenerator. I have deleted the `support_code` line here because our variables are defined globally in `objects.cu` anyways and since we have both host and device pointers, it depends on the template which variable should be use (our synapse generator template is running on the host, all others codeobjects run on the device). But I didn't quite get why cpp needs this file global variable in the first place, aren't the variables in `objects.cpp` global as well?

And we don't use the `pointers` list, we do all that in brian2cuda/device.py.

Question is: what situations am I breaking here? For the `TimedArray` implementation, I need the device pointer `d_timedarray_values` array that is global anyways, no need for the `_namespace...` version.

Final tests are running right now, but I don't think I saw any test failing because of a namespace issue.

Reply:
Just came across a documentation section on use namespace. I guess that is it. And I found the `Brian2` tests, which are not marked as `standalone-compatible` and therefore not run. Ok, I have a starting point now, I'll look into it.

Reply:
Yes, if I remember correctly this is mostly so that the values from the namespace are accessible under a known name for user-written code. And particularly the name is the same for runtime and standalone mode, so in principle you could share the C++ implementation which was relevant when we still supported `weave`.

Reply:
I played around with it and I still don't understand it. Here is an example I used with C++ standalone:

```python
set_device('cpp_standalone')

my_var = 0.1 * np.array(np.arange(5))

@implementation('cpp', '''
float foo(int i)
{
    using namespace brian;
    return my_var[i];
}''' , namespace={'my_var': my_var})
@check_units(i=1, result=1)
def foo(i):
    return my_var[i]

G = NeuronGroup(5, 'v : 1')
G.run_regularly('v = foo(i)')
net = Network(G)
net.run(defaultclock.dt)

assert_allclose(G.v_[:], my_var)
```

This generates the following code in `neurongroup_run_regularly_codeobject.cpp`:
```cpp
// support code
static double* _namespacemy_var;
float foo(int i)
{
    using namespace brian;
    return my_var[i];
}

// pointer_lines (uses namespace brian)
_namespacemy_var = my_var;
```

The file-global `_namespacemy_var` variable is not used at all, instead the global `my_var` variable is used. And the user still has to define `using namespace brian;` in their function implementation, otherwise it doesn't work.

If the goal was to make `my_var` accessible without having to add the `brian::` namespace, then the cpp function should use the `_namespacemy_var` I guess. But it doesn't.

And the only place in the brian2 code that I found where this `_namespace{var}` mechanism is actually used is in the `TimedArray` implementation, where the return value is hardcoded to be called `_namespace{name}` (see [here](https://github.com/brian-team/brian2/blob/5ed12aa774350a02699026e20227340bcab673dd/brian2/input/timedarray.py#L48)).

I think I'm still missing something here?

Reply:
Your function code gets included verbatim and does not get rewritten automatically (maybe it should?). The correct way to write your example if you follow the instructions in [the documentation](https://brian2.readthedocs.io/en/stable/advanced/functions.html#additional-namespace) is to write:
```Python
@implementation('cpp', '''
float foo(int i)
{
    return _namespacemy_var[i];
}''' , namespace={'my_var': my_var})
```
It's not really relevant anymore, but back in the days with `weave` this would have worked for both runtime and standalone mode whereas your variant hardcodes knowledge about standalone mode.

Reply:
Ahh I see. Got it now. I didn't see that part in the documentation. Alright, if it is in the docs, it should work with `_namespace{var}`. I implemented this now for brain2CUDA as well. Turns out, the mechanism is even useful for brian2CUDA, since this way the same function can be used verbatim in both, host and device code and I can just let the `_namespace{var}` pointer point to either host or device data, depending on codeobject, without the user having to care about it. Perfect.

Thanks again for your help!